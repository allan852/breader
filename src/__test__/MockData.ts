import { IntlProvider } from 'react-intl'
import { messages } from '../locales'
import { IArticle, IFeed } from '../schemas'

export const intlProviderProps: IntlProvider.Props = {
  locale: 'en-US',
  messages: messages['en-US'],
}

const intlProvider = new IntlProvider(intlProviderProps, {})
export const intl = intlProvider.getChildContext().intl

export const feed: IFeed = {
  _id: 'http://www.ruanyifeng.com/blog/atom.xml',
  _rev: '5-55772bf2b4d4b74608a60ac3480334e4',
  author: 'ruanyifeng',
  categories: [],
  createTime: 1563686468437,
  deleteTime: 0,
  description: "Ruan YiFeng's Blog",
  etag: '"11d86-58e2cc1eab031"',
  favicon: 'http://www.ruanyifeng.com/favicon.ico',
  generator: 'Movable Type 5.2.2 (http://www.sixapart.com/movabletype/)',
  language: '',
  link: 'http://www.ruanyifeng.com/blog/',
  publishTime: 1563697765000,
  time: 1563697765000,
  title: '阮一峰的网络日志',
  url: 'http://www.ruanyifeng.com/blog/atom.xml',
}

export const article: IArticle = {
  _id: 'tag:www.ruanyifeng.com,2019:/blog//1.2100',
  _rev: '5-ff6fc450616071a8f7d022b7532e85aa',
  author: 'ruanyifeng',
  categories: [],
  comments: 'article comments',
  createTime: 1563697765000,
  deleteTime: 0,
  description: 'article description',
  enclosures: [],
  feedId: 'http://www.ruanyifeng.com/blog/atom.xml',
  image: '',
  isDayFirst: true,
  isStarred: false,
  isUnread: false,
  link: 'http://www.ruanyifeng.com/blog/2019/07/weekly-issue-65.html',
  originLink: '',
  publishTime: 1563697765000,
  summary: 'article summary',
  time: 1563697765000,
  title: 'article title',
}
