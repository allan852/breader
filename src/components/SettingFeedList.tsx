import { Avatar, Button, List as AntdList } from 'antd'
import React, { PureComponent } from 'react'
import { Scrollbars } from 'react-custom-scrollbars'
import { FormattedMessage, injectIntl, WrappedComponentProps } from 'react-intl'
import { IFeed } from '../schemas'
import '../styles/SettingFeedList.less'
import Utils from '../utils'

export interface ISettingFeedListProps {
  feeds: IFeed[]
  onDeleteFeed?: (feedId: string, feedIndex: number) => void
}
export interface ISettingFeedListState {}

export class SettingFeedListComponent extends PureComponent<
  ISettingFeedListProps & WrappedComponentProps,
  ISettingFeedListState
> {
  public constructor(props: ISettingFeedListProps & WrappedComponentProps) {
    super(props)
  }
  public handleDeleteClick = (feedId: string, feedIndex: number) => {
    const { onDeleteFeed } = this.props
    if (onDeleteFeed) {
      onDeleteFeed(feedId, feedIndex)
    }
  }
  public handleScrollStart() {
    // fix: scrollbar not get close bottom
    const uls = document.getElementsByClassName('setting-feed-items')
    const ul = uls.item(0) as HTMLElement
    if (ul) {
      ul.style.paddingLeft = '0.5px'
      setImmediate(() => {
        ul.style.paddingLeft = null
      })
    }
  }
  public render() {
    const { feeds } = this.props
    return (
      <div className="ant-list ant-list-sm ant-list-split ant-list-bordered setting-feed-list">
        <Scrollbars
          autoHide
          autoHeight
          autoHeightMax={`calc(${Utils.getClientHightForCalc()} - 332px)`}
          onScrollStart={this.handleScrollStart}>
          <ul className="ant-list-items setting-feed-items">
            {feeds.map((feed, index) => (
              <AntdList.Item
                className="setting-feed-item"
                key={feed._id}
                actions={[
                  <Button
                    key={feed._id}
                    size="small"
                    type="danger"
                    onClick={() => this.handleDeleteClick(feed._id, index)}>
                    <FormattedMessage id="delete" />
                  </Button>,
                ]}>
                <p title={feed.url} className="feed-item-content">
                  <Avatar shape="square" size={16} src={feed.favicon} />{' '}
                  {feed.title}
                </p>
              </AntdList.Item>
            ))}
          </ul>
        </Scrollbars>
        {!feeds.length && (
          <p className="empty-list">
            <FormattedMessage id="noFeeds" />
          </p>
        )}
      </div>
    )
  }
}

export default injectIntl(SettingFeedListComponent)
