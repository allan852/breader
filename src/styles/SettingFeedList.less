@import './Values.less';
.setting-feed-list {
  .feed-item-content {
    flex: 1;
    margin-bottom: 0;
  }
  .empty-list {
    color: @black-text-color-dark;
    height: 40px;
    padding: 8px 0px;
    text-align: center;
    line-height: 21px;
    margin-bottom: 0;
  }
}
