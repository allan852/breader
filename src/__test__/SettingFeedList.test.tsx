import Enzyme from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import EnzymeToJson from 'enzyme-to-json'
import React from 'react'
import { Scrollbars } from 'react-custom-scrollbars'
import { IntlProvider, WrappedComponentProps } from 'react-intl'
import SettingFeedList, {
  ISettingFeedListProps,
  ISettingFeedListState,
  SettingFeedListComponent,
} from '../components/SettingFeedList'
import { domNode, feed, intlProviderProps } from './MockData'

Enzyme.configure({ adapter: new EnzymeAdapter() })

describe('SettingFeedList Testing', () => {
  jest.useFakeTimers()
  const mockProps: ISettingFeedListProps = {
    feeds: [feed],
    onDeleteFeed: jest.fn(),
  }

  it('dom testing', () => {
    const wrapper = Enzyme.mount(
      <IntlProvider {...intlProviderProps}>
        <SettingFeedList {...mockProps} />
      </IntlProvider>,
      { attachTo: domNode }
    )
    wrapper.update()
    // const sbar = wrapper.find(Scrollbars) as Enzyme.ReactWrapper<ScrollbarProps, {}, Scrollbars>
    // sbar.instance().scrollToTop()
    const feedList = wrapper.find(
      SettingFeedListComponent
    ) as Enzyme.ReactWrapper<
      ISettingFeedListProps & WrappedComponentProps,
      ISettingFeedListState,
      SettingFeedListComponent
    >
    feedList.instance().handleScrollStart()
    wrapper.find('.ant-btn-danger').simulate('click')
    expect(mockProps.onDeleteFeed).toBeCalledTimes(1)
    expect(wrapper.find(Scrollbars).find('li.setting-feed-item')).toHaveLength(
      1
    )
  })

  it('snapshot testing', () => {
    const component = Enzyme.mount(
      <IntlProvider {...intlProviderProps}>
        <SettingFeedList {...mockProps} />
      </IntlProvider>
    )
    const tree = EnzymeToJson(component)
    expect(tree).toMatchSnapshot()
  })
})
