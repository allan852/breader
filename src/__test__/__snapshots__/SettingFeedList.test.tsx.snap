// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingFeedList Testing snapshot testing 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en-US"
  messages={
    Object {
      "addFeed": "Add Feed",
      "allArticlesAreReadNow": "All articles are read now.",
      "delete": "Delete",
      "doYouWantSetAllArticlesBeRead": "Do you want set all articles be read?",
      "feedUrl": "Feed Url",
      "feeds": "Feeds",
      "feedsAreUpdated": "Feeds are Updated",
      "inputKeywords": "Please input the keywords",
      "invalidFeedUrl": "Invalid Feed Url",
      "keywords": "Keywords",
      "languages": "Languages",
      "menuAllItems": "All Items",
      "menuStarred": "Starred",
      "menuSubscriptions": "Subscriptions",
      "menuUnread": "Unread",
      "noFeeds": "No Feeds",
      "noMatched": "No Matched",
      "settings": "Settings",
      "somethingWrong": "Something Wrong!",
      "title": "Breader",
      "unfoundFeed": "Unfound Feed",
      "unknown": "Unknown",
    }
  }
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <injectIntl(SettingFeedListComponent)
    feeds={
      Array [
        Object {
          "_id": "http://www.ruanyifeng.com/blog/atom.xml",
          "_rev": "5-55772bf2b4d4b74608a60ac3480334e4",
          "author": "ruanyifeng",
          "categories": Array [],
          "createTime": 1563686468437,
          "deleteTime": 0,
          "description": "Ruan YiFeng's Blog",
          "etag": "\\"11d86-58e2cc1eab031\\"",
          "favicon": "http://www.ruanyifeng.com/favicon.ico",
          "generator": "Movable Type 5.2.2 (http://www.sixapart.com/movabletype/)",
          "language": "",
          "link": "http://www.ruanyifeng.com/blog/",
          "publishTime": 1563697765000,
          "time": 1563697765000,
          "title": "阮一峰的网络日志",
          "url": "http://www.ruanyifeng.com/blog/atom.xml",
        },
      ]
    }
    onDeleteFeed={
      [MockFunction] {
        "calls": Array [
          Array [
            "http://www.ruanyifeng.com/blog/atom.xml",
            0,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <SettingFeedListComponent
      feeds={
        Array [
          Object {
            "_id": "http://www.ruanyifeng.com/blog/atom.xml",
            "_rev": "5-55772bf2b4d4b74608a60ac3480334e4",
            "author": "ruanyifeng",
            "categories": Array [],
            "createTime": 1563686468437,
            "deleteTime": 0,
            "description": "Ruan YiFeng's Blog",
            "etag": "\\"11d86-58e2cc1eab031\\"",
            "favicon": "http://www.ruanyifeng.com/favicon.ico",
            "generator": "Movable Type 5.2.2 (http://www.sixapart.com/movabletype/)",
            "language": "",
            "link": "http://www.ruanyifeng.com/blog/",
            "publishTime": 1563697765000,
            "time": 1563697765000,
            "title": "阮一峰的网络日志",
            "url": "http://www.ruanyifeng.com/blog/atom.xml",
          },
        ]
      }
      intl={
        Object {
          "defaultFormats": Object {},
          "defaultLocale": "en",
          "formatDate": [Function],
          "formatDateToParts": [Function],
          "formatHTMLMessage": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatNumberToParts": [Function],
          "formatPlural": [Function],
          "formatRelativeTime": [Function],
          "formatTime": [Function],
          "formatTimeToParts": [Function],
          "formats": Object {},
          "formatters": Object {
            "getDateTimeFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralRules": [Function],
            "getRelativeTimeFormat": [Function],
          },
          "locale": "en-US",
          "messages": Object {
            "addFeed": "Add Feed",
            "allArticlesAreReadNow": "All articles are read now.",
            "delete": "Delete",
            "doYouWantSetAllArticlesBeRead": "Do you want set all articles be read?",
            "feedUrl": "Feed Url",
            "feeds": "Feeds",
            "feedsAreUpdated": "Feeds are Updated",
            "inputKeywords": "Please input the keywords",
            "invalidFeedUrl": "Invalid Feed Url",
            "keywords": "Keywords",
            "languages": "Languages",
            "menuAllItems": "All Items",
            "menuStarred": "Starred",
            "menuSubscriptions": "Subscriptions",
            "menuUnread": "Unread",
            "noFeeds": "No Feeds",
            "noMatched": "No Matched",
            "settings": "Settings",
            "somethingWrong": "Something Wrong!",
            "title": "Breader",
            "unfoundFeed": "Unfound Feed",
            "unknown": "Unknown",
          },
          "onError": [Function],
          "textComponent": Symbol(react.fragment),
          "timeZone": undefined,
        }
      }
      onDeleteFeed={
        [MockFunction] {
          "calls": Array [
            Array [
              "http://www.ruanyifeng.com/blog/atom.xml",
              0,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    >
      <div
        className="ant-list ant-list-sm ant-list-split ant-list-bordered setting-feed-list"
      >
        <Scrollbars
          autoHeight={true}
          autoHeightMax="calc(100vh - 332px)"
          autoHeightMin={0}
          autoHide={true}
          autoHideDuration={200}
          autoHideTimeout={1000}
          hideTracksWhenNotNeeded={false}
          onScrollStart={[Function]}
          renderThumbHorizontal={[Function]}
          renderThumbVertical={[Function]}
          renderTrackHorizontal={[Function]}
          renderTrackVertical={[Function]}
          renderView={[Function]}
          tagName="div"
          thumbMinSize={30}
          universal={false}
        >
          <div
            style={
              Object {
                "height": "auto",
                "maxHeight": "calc(100vh - 332px)",
                "minHeight": 0,
                "overflow": "hidden",
                "position": "relative",
                "width": "100%",
              }
            }
          >
            <div
              key="view"
              style={
                Object {
                  "WebkitOverflowScrolling": "touch",
                  "bottom": undefined,
                  "left": undefined,
                  "marginBottom": 0,
                  "marginRight": 0,
                  "maxHeight": "calc(calc(100vh - 332px) + 0px)",
                  "minHeight": 0,
                  "overflow": "scroll",
                  "position": "relative",
                  "right": undefined,
                  "top": undefined,
                }
              }
            >
              <ul
                className="ant-list-items setting-feed-items"
              >
                <Item
                  actions={
                    Array [
                      <Button
                        block={false}
                        ghost={false}
                        htmlType="button"
                        loading={false}
                        onClick={[Function]}
                        size="small"
                        type="danger"
                      >
                        <FormattedMessage
                          id="delete"
                          values={Object {}}
                        />
                      </Button>,
                    ]
                  }
                  className="setting-feed-item"
                  key="http://www.ruanyifeng.com/blog/atom.xml"
                >
                  <li
                    className="ant-list-item setting-feed-item"
                  >
                    <p
                      className="feed-item-content"
                      title="http://www.ruanyifeng.com/blog/atom.xml"
                    >
                      <Avatar
                        shape="square"
                        size={16}
                        src="http://www.ruanyifeng.com/favicon.ico"
                      >
                        <span
                          className="ant-avatar ant-avatar-square ant-avatar-image"
                          style={
                            Object {
                              "fontSize": 18,
                              "height": 16,
                              "lineHeight": "16px",
                              "width": 16,
                            }
                          }
                        >
                          <img
                            onError={[Function]}
                            src="http://www.ruanyifeng.com/favicon.ico"
                          />
                        </span>
                      </Avatar>
                       
                      阮一峰的网络日志
                    </p>
                    <ul
                      className="ant-list-item-action"
                      key="actions"
                    >
                      <li
                        key="ant-list-item-action-0"
                      >
                        <Button
                          block={false}
                          ghost={false}
                          htmlType="button"
                          key="http://www.ruanyifeng.com/blog/atom.xml"
                          loading={false}
                          onClick={[Function]}
                          size="small"
                          type="danger"
                        >
                          <Wave>
                            <button
                              className="ant-btn ant-btn-danger ant-btn-sm"
                              onClick={[Function]}
                              type="button"
                            >
                              <FormattedMessage
                                id="delete"
                                key=".0"
                                values={Object {}}
                              >
                                Delete
                              </FormattedMessage>
                            </button>
                          </Wave>
                        </Button>
                      </li>
                    </ul>
                  </li>
                </Item>
              </ul>
            </div>
            <div
              key="trackHorizontal"
              style={
                Object {
                  "borderRadius": 3,
                  "bottom": 2,
                  "display": "none",
                  "height": 6,
                  "left": 2,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 2,
                  "transition": "opacity 200ms",
                }
              }
            >
              <div
                style={
                  Object {
                    "backgroundColor": "rgba(0,0,0,.2)",
                    "borderRadius": "inherit",
                    "cursor": "pointer",
                    "display": "block",
                    "height": "100%",
                    "position": "relative",
                  }
                }
              />
            </div>
            <div
              key="trackVertical"
              style={
                Object {
                  "borderRadius": 3,
                  "bottom": 2,
                  "display": "none",
                  "opacity": 0,
                  "position": "absolute",
                  "right": 2,
                  "top": 2,
                  "transition": "opacity 200ms",
                  "width": 6,
                }
              }
            >
              <div
                style={
                  Object {
                    "backgroundColor": "rgba(0,0,0,.2)",
                    "borderRadius": "inherit",
                    "cursor": "pointer",
                    "display": "block",
                    "position": "relative",
                    "width": "100%",
                  }
                }
              />
            </div>
          </div>
        </Scrollbars>
      </div>
    </SettingFeedListComponent>
  </injectIntl(SettingFeedListComponent)>
</IntlProvider>
`;
